unblob/handlers/filesystem/fat.py

```diff
@@ -155,6 +155,11 @@ def calculate_chunk(self, file: File, start_offset: int) -> Optional[ValidChunk]
 
         size = header.common.BytesPerSec * sector_count
 
+        # handle corrupted FAT images by considering the whole file as the FAT image.
+        file_size = file.size()
+        if start_offset == 0 and size > file_size:
```

Comment:
do we know the overflowed chunk size? maybe we can consider that

Reply:
Looking at the Android OTA sample, we have these FAT16 images:

```
bluetooth.img:         DOS/MBR boot sector, code offset 0x3c+2, OEM-ID "MSDOS5.0", Bytes/sector 4096, sectors/cluster 4, root entries 512, Media descriptor 0xf8, sectors/FAT 3, sectors/track 63, heads 255, sectors 16384 (volumes > 32 MB), serial number 0xbc614e, unlabeled, FAT (16 bit)
modem_apq.img:         DOS/MBR boot sector, code offset 0x3c+2, OEM-ID "MSDOS5.0", sectors/cluster 32, root entries 512, Media descriptor 0xf8, sectors/FAT 28, sectors/track 63, heads 255, sectors 225280 (volumes > 32 MB), serial number 0xbc614e, unlabeled, FAT (16 bit)
modem.img:             DOS/MBR boot sector, code offset 0x3c+2, OEM-ID "MSDOS5.0", sectors/cluster 32, root entries 512, Media descriptor 0xf8, sectors/FAT 28, sectors/track 63, heads 255, sectors 225280 (volumes > 32 MB), serial number 0xbc614e, unlabeled, FAT (16 bit)
```

Looking at them with `7z`:

```
Path = bluetooth.img
Type = FAT
ERRORS:
Unexpected end of archive
Physical Size = 67108864
File System = FAT16
Cluster Size = 16384
Free Space = 66338816
Headers Size = 61440
Sector Size = 4096
ID = 12345678


Path = modem_apq.img
Type = FAT
ERRORS:
Unexpected end of archive
Physical Size = 115343360
File System = FAT16
Cluster Size = 16384
Free Space = 73269248
Headers Size = 160256
Sector Size = 512
ID = 12345678


Path = modem.img
Type = FAT
ERRORS:
Unexpected end of archive
Physical Size = 115343360
File System = FAT16
Cluster Size = 16384
Free Space = 3981312
Headers Size = 5714432
Sector Size = 512
ID = 12345678
```

So in terms of difference between size in header and actual size:

```
67108864 - 765952 = 66342912
115343360 - 42074112 = 73269248
115343360 - 111362048 = 3981312
```

In terms of ratio of actual size vs size stored in header:

```
>>> (765952 / 67108864) * 100
1.141357421875
>>> (42074112 / 115343360) * 100
36.47727272727273
>>> (111362048 / 115343360) * 100
96.54829545454545
```

Reply:
hmm, it looks like that the actual size is around the "Physical size" - "Free Space", at least for the modem_apq.img and modem.img and for bluetooth.img it is almost the same except off by one sector which is 4k in that case. 

Maybe we should use this as a fall-back calculation?

Reply:
We need to implement a file allocation table parser to calculate the free space. It's not trivial but we'll get an enhanced chunk size calculation out of it. I'll look into it.

Reply:
Pushed a new solution that fallback to FAT parsing to calculate the end offset. Everything is explained in the updated description.